src/vterm.c

```diff
@@ -0,0 +1,474 @@
+
+#include "vterm.h"
+
+#include "lisp.h"
+#include "coding.h"
+
+
+/* #define SB_MAX 100000 // Maximum 'scrollback' value. */
+
+#ifndef MIN
+#define MIN(X, Y) ((X) < (Y) ? (X) : (Y))
+#endif
+#ifndef MAX
+#define MAX(X, Y) ((X) > (Y) ? (X) : (Y))
+#endif
+
+Term *global_term;
+
+static void fetch_cell(Term *, int , int , VTermScreenCell *);
+static bool compare_cells(VTermScreenCell *, VTermScreenCell *);
+static bool is_key(unsigned char *key, size_t len, const char *key_description);
+
+VTermColor rgb_string_to_color(Lisp_Object string);
+
+
+static int term_sb_push(int cols, const VTermScreenCell *cells, void *data) {
+  Term *term = (Term *)data;
+
+  if (!term->sb_size) {
+    return 0;
+  }
+
+  // copy vterm cells into sb_buffer
+  size_t c = (size_t)cols;
+  VtermScrollbackLine *sbrow = NULL;
+  if (term->sb_current == term->sb_size) {
+    if (term->sb_buffer[term->sb_current - 1]->cols == c) {
+      // Recycle old row if it's the right size
+      sbrow = term->sb_buffer[term->sb_current - 1];
+    } else {
+      free(term->sb_buffer[term->sb_current - 1]);
+    }
+
+    // Make room at the start by shifting to the right.
+    memmove(term->sb_buffer + 1, term->sb_buffer,
+            sizeof(term->sb_buffer[0]) * (term->sb_current - 1));
+
+  } else if (term->sb_current > 0) {
+    // Make room at the start by shifting to the right.
+    memmove(term->sb_buffer + 1, term->sb_buffer,
+            sizeof(term->sb_buffer[0]) * term->sb_current);
+  }
+
+  if (!sbrow) {
+    sbrow = malloc(sizeof(VtermScrollbackLine) + c * sizeof(sbrow->cells[0]));
+    sbrow->cols = c;
+  }
+
+  // New row is added at the start of the storage buffer.
+  term->sb_buffer[0] = sbrow;
+  if (term->sb_current < term->sb_size) {
+    term->sb_current++;
+  }
+
+  if (term->sb_pending < (int)term->sb_size) {
+    term->sb_pending++;
+  }
+
+  memcpy(sbrow->cells, cells, sizeof(cells[0]) * c);
+
+  return 1;
+}
+
+/// Scrollback pop handler (from pangoterm).
+///
+/// @param cols
+/// @param cells  VTerm state to update.
+/// @param data   Term
+static int term_sb_pop(int cols, VTermScreenCell *cells, void *data) {
+  Term *term = (Term *)data;
+
+  if (!term->sb_current) {
+    return 0;
+  }
+
+  if (term->sb_pending) {
+    term->sb_pending--;
+  }
+
+  VtermScrollbackLine *sbrow = term->sb_buffer[0];
+  term->sb_current--;
+  // Forget the "popped" row by shifting the rest onto it.
+  memmove(term->sb_buffer, term->sb_buffer + 1,
+          sizeof(term->sb_buffer[0]) * (term->sb_current));
+
+  size_t cols_to_copy = (size_t)cols;
+  if (cols_to_copy > sbrow->cols) {
+    cols_to_copy = sbrow->cols;
+  }
+
+  // copy to vterm state
+  memcpy(cells, sbrow->cells, sizeof(cells[0]) * cols_to_copy);
+  size_t col;
+  for (col = cols_to_copy; col < (size_t)cols; col++) {
+    cells[col].chars[0] = 0;
+    cells[col].width = 1;
+  }
+
+  free(sbrow);
+
+  return 1;
+}
+
+static size_t
+codepoint_to_utf8_c(const uint32_t codepoint, unsigned char buffer[4]) {
```

Comment:
This (and the subsequent `utf8_to_codepoint`) are very similar to functions that already exist in `libvterm`. It might be easier just to expose those publicly then you can use them directly.

Reply:
Good to know. I will make a TODO list.